### PR TITLE
Added $skin option in AssetBundle and created Helper class

### PR DESCRIPTION
**Pull request by freezy-sk**
Added $skin option in AssetBundle and created Helper for selected skin class

### CODE REVIEW

**Comment by schmunk42** [comment]
You can already use skins, just by specifying the class for the body-tag.

```
<body class="skin-black">
```

IMHO a much slimmer solution would be just to use a `param`, but please let me know if I've overlooked something here.

**Comment by freezy-sk** [comment]
@schmunk42 current implementation is loading css with all skins which is about 42KB minified and can be much bigger when they add more colors. My small change is loading just 4KB skin file and is not dependent on application params.

**Comment by schmunk42** [comment]
@freezy-sk 
- Does this approach still works when bundling assets? Would be nice, if we could choose from bundling all assets or just the one specified in the config.
- How about using constants for the skin names in the asset bundle, this way we could skip the checks and IMHO we also don't need the extra helper class then.

**Comment by freezy-sk** [comment]
@schmunk42
* Right now AssetBundle just copy everything what is under `$sourcePath` into public assets directory. These files can be filtered, but I think it's not necessary because what really depends is what is downloaded to users (defined in $css and $js).
* I think that using constants for skins will just lead to loss of flexibility and not really solve the problem. See examples below.

**String option and helper class**
```php
'assetManager' => [
    'bundles' => [
        'dmstr\web\AdminLteAsset' => [
            'sourcePath' => '@bower/admin-lte/dist',
            'skin' => 'skin-black',
        ],
    ],
],

<body class="<?= \dmstr\helpers\AdminLteHelper::skinClass() ?>">
```

**Skin constants**
```php
'assetManager' => [
    'bundles' => [
        'dmstr\web\AdminLteAsset' => [
            'sourcePath' => '@bower/admin-lte/dist',
            'skin' => \dmstr\web\AdminLteAsset::SKIN_BLUE,
        ],
    ],
],

<body class="<?= \dmstr\web\AdminLteAsset::SKIN_BLUE ?>">
```

**Comment by schmunk42** [comment]
Source path is the `dist` folder:

    public $sourcePath = '@bower/admin-lte/dist';

So copying 40kB into assets is not the big deal. But you're right about serving the files to the end user. It would be nicer, if it was smaller.

My point was about *bundling and compression* of assets. I think your approach wouldn't work 100% with it, because the asset-skin is determined at runtime, so there's no possibility for Yii2 to detect which skin it has to bundle.
That was what I meant, when saying this should be an optional setting, maybe we should introduce two asset bundles here: One with all skins (like the current one) and one without any skins.
You'd be still able to bundle and compress, but your additional skin file would be served separately. AFAIK Yii2 just reads the public properties of an asset bundle to determine which files can be bundled, see [Combining and Compressing Assets](http://www.yiiframework.com/doc-2.0/guide-structure-assets.html#combining-and-compressing-assets).

And fore sure, if there'll be more colors, we'd have to add more constants.

**Comment by freezy-sk** [comment]
@schmunk42 I think it will work because bundles are accessed instantiated so there will be correct list of css files.

**Comment by Faryshta** [comment]
https://github.com/yiisoft/yii2-jui/blob/master/DatePickerLanguageAsset.php

For the datepicker language a similar solution is implemented but not the same exactly.

Also I don't think its ok to set the skin on the config since it would mean i will have to use the same skin on all my pages, sometimes i want to show different skins to different users or the option to change skins.

My proposal is to build an `Skin` widget.

Example

```php
$container = AdminLteContainer::begin([
   'skin' => 'light-blue',
   'tag' => 'body'
   'options' => 
]);

echo $container->beginHeader();

echo $container->endHeader();

echo $container->beginSidebar();

echo $container->endSidebar();

echo $container->beginContent();

echo $content;

echo $container->endContent();

AdminLteContainer::end();
```

**Comment by Faryshta** [comment]
@freezy-sk @schmunk42 do you like the widget idea?

**Comment by freezy-sk** [comment]
@Faryshta I thing this is complicated solution. Better will be making some complete views which can be used as theme views should be. http://www.yiiframework.com/doc-2.0/guide-output-theming.html

**Comment by Faryshta** [comment]
i think that doesn't solve the skin issue. A theme can use one, many or dynamic skins. Actually i am working on trying to use dynamic skins for my theme.

We will still need  a form for the asset to dynamically know which skin to load.

**Comment by schmunk42** [comment]
Thanks! 

I think this is the right approach, although I may need to check asset bundling again.

**Comment by Faryshta** [comment]
I still have the same question. How can i load the skins dynamically?

**Comment by schmunk42** [comment]
How about an option to switch the behavior?
Could still be made backward compatible.

**Comment by freezy-sk** [comment]
@Faryshta You can just ignore `'skin'` option and load your own css files
```php
'components' => [
    'assetManager' => [
        'bundles' => [
            'dmstr\web\AdminLteAsset' => [
                'css' => [
                    'css/AdminLTE.min.css',
                    'css/skins/_all-skins.min.css',
                ],
            ],
        ],
    ],
],
```

**Comment by schmunk42** [comment]
@Faryshta @freezy-sk Please have a look:
https://github.com/dmstr/yii2-adminlte-asset/commit/2694edd7dacd9204820c8e0c21bd07d816735310

Should be backward-compatible now ... and by setting the property you'll register the CSS file and can use the helper then.

**Comment by freezy-sk** [comment]
:+1:

**Comment by Faryshta** [comment]
how would that behavior be?

**Comment by Faryshta** [comment]
i think i haven't explained myself properly.

Lets say I have two type of users for my admin.

salesman and manager

they both see different menus when they start session, they see different options and have different permissions based on their role.

Now i want that the salesman have an skin-blue when they login and the mangers an skin-red. Mostly to make it more obvious which user they are using at the moment (specially useful when your store doesn't have many computers and salesman or manager have to share computer for a few times)

Or let my users choose the skin they prefer (specially useful if two salesman have to use the same computer)

how can that be achieved?

**Comment by freezy-sk** [comment]
@Faryshta you can do this by some custom helper class which will give you the right skin class to your `<body>` based on your user's role. You just need to have loaded `_all-skins.css` which would be still default css when you don't change skin option from default.

**Comment by Faryshta** [comment]
Then we need to make a documented way to do it properly, i still think the widget is the best option we have. maybe my example was too bulky, just `AdminLteContainer::widget(['skin' => 'light-blue'])` can be a better idea.

and not just skin can get customized using the widget

**Comment by githubjeka** [comment]
Helper class return `'_all-skins'` css of body class >_<

File: helpers/AdminLteHelper.php [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+class AdminLteHelper
+{
+    /**
+     * @return string
+     */
+    public static function skinClass()
+    {
+        /** @var \dmstr\web\AdminLteAsset $bundle */
+        $bundle = Yii::$app->assetManager->getBundle('dmstr\web\AdminLteAsset');
```
Comment (githubjeka) [inline helpers/AdminLteHelper.php]: here

File: web/AdminLteAsset.php [hunk 2]
```diff
@@ -22,4 +22,27 @@ class AdminLteAsset extends AssetBundle
...
+
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        // Append skin color file if specified
+        if ($this->skin) {
+            if (strpos($this->skin, 'skin-') !== 0) {
+                throw new Exception('Invalid skin specified');
```
Comment (githubjeka) [inline web/AdminLteAsset.php]: Exception is undefined

**Comment by schmunk42** [comment]
I can't recall it 100%, but I think I had to made it that way to keep BC.
Could you open a new issue ... we should also add some tests ;) Will try to come up with a neat solution.

**Pull request merged**